src/dispatch/batch_builder.rs

```diff
@@ -0,0 +1,496 @@
+        self.stages_builder.add_barrier();
+    }
+
+    fn next_id(&mut self) -> SystemId {
+        let id = self.current_id;
+        self.current_id += 1;
+
+        SystemId(id)
+    }
+
+    /// Build the BatchSystem which take care to dispatch the subsystems.
+    pub(crate) fn build(mut self) -> (Vec<ResourceId>, Vec<ResourceId>, BatchSystem<'a>) {
+        self.read_resources.sort();
+        self.read_resources.dedup();
+
+        self.write_resources.sort();
+        self.write_resources.dedup();
+
+        let stage = self.stages_builder.build();
+
+        // Average
+        let batch_running_time = self.running_time / (self.system_count as f32);
+        // Depending on the system count increase the running time
+        let batch_running_time = batch_running_time + 2.0 * (self.system_count as f32 / 15.0);
+        let batch_running_time = batch_running_time.round() as i32;
+
+        let batch_running_time = match batch_running_time {
+            1 => RunningTime::VeryShort,
+            2 => RunningTime::Short,
+            3 => RunningTime::Average,
+            4 => RunningTime::Long,
+            _ => RunningTime::VeryLong,
+        };
+
+        #[cfg(feature = "parallel")]
+        let res = (
+            self.read_resources,
+            self.write_resources,
+            BatchSystem::new(
+                stage,
+                self.batch_controller,
+                batch_running_time,
+                self.thread_pool.unwrap(),
+            ),
+        );
+
+        #[cfg(not(feature = "parallel"))]
+        let res = (
+            self.read_resources,
+            self.write_resources,
+            Batch::new(stage, batch_running_time, batch_running_time),
+        );
+
+        res
+    }
+}
+
+pub struct UncheckedWorld<'a>(pub &'a World);
+
+impl<'a> SystemData<'a> for UncheckedWorld<'a> {
+    fn setup(_world: &mut World) {}
+
+    /// Returns the world
+    fn fetch(res: &'a World) -> Self {
+        UncheckedWorld(res)
+    }
+
+    /// Doesn't return nothing.
+    /// To use this Struct is necessary check the dependecy manually
+    fn reads() -> Vec<ResourceId> {
+        vec![]
+    }
+
+    /// Doesn't return nothing.
+    /// To use this Struct is necessary check the dependecy manually
+    fn writes() -> Vec<ResourceId> {
+        vec![]
+    }
+}
+
+/// The `BatchSystem` allow to register some subsystems that it will take
+/// care to execute once the main dispatcher execute it.
+/// It has the capability to restart its systems many times, and this
+/// can be controlled by the user.
+///
+/// The BatchSystem keep track of its internal dependecies and is able to run
+/// in parallel with other systems.
+/// At the same time the subsystems can run in parallel within the BatchSystem.
+///
+/// To add a new batch system you have to call the function `with_batch` which
+/// accept a `BatchBuilder` object that will take care to construct the `BatchSystem`.
+pub struct BatchSystem<'a> {
```

Comment:
Do we even need this? This looks exactly like `Dispatcher`, with the exception of `batch_controller`, which I don't think should be in here.